### PR TITLE
Add an Elasticsearch store interface

### PR DESCRIPTION
**Pull request by jessemyers**
Includes:
 -  A model base class with common attributes
 -  A store abstraction that parallels `microcosm-postgres.store.Store`
 -  Preliminary index management support
 -  Better unit test support in ES factoriews
 -  Example test fixtures

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,58 +7,41 @@ Elasticsearch configuration using [microcosm](https://github.com/globality-corp/
 
 ## Usage
 
-    from microcosm.api import create_object_graph
```
Comment (jessemyers) [inline README.md]: This example code didn't feel relevant to anything we've done.

File: README.md [hunk 2]
```diff
@@ -7,58 +7,41 @@ Elasticsearch configuration using [microcosm](https://github.com/globality-corp/
...
+
+         nosetests
 
 
 ## Configuration
 
 When using with an AWS Elasticsearch instance, use:
 
-    config.elasticsearch_client.use_aws4auth = True
+    config.elasticsearch_client.use_aws4auth = 'true'
```
Comment (jessemyers) [inline README.md]: Configuration parameters now assume strings to support environment variables properly.

File: microcosm_elasticsearch/factories.py [hunk 3]
```diff
@@ -23,7 +23,7 @@
     password=None,
     use_aws4auth="false",
     use_aws_instance_metadata="false",
-    use_python2_serializer="true",
+    use_python2_serializer="false",
```
Comment (jessemyers) [inline microcosm_elasticsearch/factories.py]: Assuming Python3 by default now.

File: microcosm_elasticsearch/factories.py [hunk 4]
```diff
@@ -32,26 +32,44 @@ def configure_elasticsearch_client(graph):
...
+    )
+
+    if graph.config.elasticsearch_client.username and graph.config.elasticsearch_client.password:
+        config.update(
+            http_auth=(
+                graph.config.elasticsearch_client.username,
+                graph.config.elasticsearch_client.password,
+            ),
+        )
+    elif graph.metadata.testing:
```
Comment (jessemyers) [inline microcosm_elasticsearch/factories.py]: This is new (default unit test credentials); the rest is just refactoring for clarity

File: microcosm_elasticsearch/indexing.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+"""
+Index management.
```
Comment (jessemyers) [inline microcosm_elasticsearch/indexing.py]: I still need to research our and the communities migration approaches; I'm hoping to grow this module into a proper set of tools for managing indexes, including migrations.

File: microcosm_elasticsearch/store.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
+"""
+Abstraction layer for persistence operations.
+
+Allows upstream (e.g. HTTP) operations to use a consistent persistence interface
+and avoids leaking persistence abstractions throughout a code base.
+
+Intended to be duck-type compatible with `microcosm_postgres.store.Store`.
```
Comment (jessemyers) [inline microcosm_elasticsearch/store.py]: Big win here!

File: microcosm_elasticsearch/store.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+class Store(object):
+    """
+    Elasticsearch persistence interface.
+
+    """
+    def __init__(self, graph, index, doc_type):
+        """
+        :param graph: the object graph
+        :param index: the name of an index to use
+        :param doc_type: a `elasticsearch_dsl.DocType` subclass to persist.
```
Comment (jessemyers) [inline microcosm_elasticsearch/store.py]: I'm using `elasticsearch_dsl` as a pseudo ORM. Most of the "interesting" work I did in my first draft (API calls, query abstraction) are already part of this DSL. It's not perfect, but it's certainly good enough.

File: microcosm_elasticsearch/tests/fixtures.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    middle = Text(required=False)
+    last = Text(required=True)
+
+
+@binding("person_store")
+class PersonStore(Store):
+    def __init__(self, graph):
+        super(PersonStore, self).__init__(graph, graph.example_index, Person)
+
+    def _filter(self, query, q=None, **kwargs):
```
Comment (jessemyers) [inline microcosm_elasticsearch/tests/fixtures.py]: Here's an example of customizing the `Store.search` to do something interesting. We can obviously do a lot more.

File: microcosm_elasticsearch/tests/fixtures.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+"""
+from elasticsearch_dsl import Q, Text
+from microcosm.api import binding
+
+from microcosm_elasticsearch.indexing import make_index
+from microcosm_elasticsearch.models import Model
+from microcosm_elasticsearch.store import Store
+
+
+@binding("example_index")
```
Comment (jessemyers) [inline microcosm_elasticsearch/tests/fixtures.py]: We always need to choose an index.

I'm assuming a microservice will have a few, static indexes right now, but we may end up with a broader index management layer eventually. In this case, we'll need some sort of delegation layer that lets a store find its current/correct index.

File: microcosm_elasticsearch/tests/fixtures.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+def create_example_index(graph):
+    return make_index(graph)
+
+
+class Person(Model):
```
Comment (jessemyers) [inline microcosm_elasticsearch/tests/fixtures.py]: FYI: `String` is deprecated; use `Text` or `Keyword`. Also `UUID` isn't supported really.

File: microcosm_elasticsearch/tests/test_store.py [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+        self.store.create(self.kevin)
+        assert_that(
+            calling(self.store.create).with_args(self.kevin),
+            raises(ElasticsearchConflictError),
+        )
+
+    def test_count(self):
+        self.store.create(self.kevin)
+        self.store.create(self.steph)
+        assert_that_eventually(
```
Comment (jessemyers) [inline microcosm_elasticsearch/tests/test_store.py]: `count` (and `search`) are queries and will not immediately read our writes.
`get` and `delete` do not have delays.

File: microcosm_elasticsearch/store.py [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+
+    def _filter(self, query, offset=None, limit=None, **kwargs):
+        """
+        Filter a query with user-supplied arguments.
+
+        :param offset: pagination offset, if any
+        :param limit: pagination limit, if any
+
+        """
+        if offset is not None and limit is not None:
```
Comment (netama) [inline microcosm_elasticsearch/store.py]: should we require both together?

**Comment by netama** [comment]
LGTM with one small question. Cool :)

File: microcosm_elasticsearch/store.py (hunk 12 above)
Comment (jessemyers) [inline microcosm_elasticsearch/store.py]: what's the use case for these separately? just using the default values?

File: microcosm_elasticsearch/indexing.py (hunk 5 above)
Comment (cdm-ium) [inline microcosm_elasticsearch/indexing.py]: I would argue the most basic, useful things you could add, which we have in gaston today are:

- a cli for iterating over documents which match some query. We have a cli for iterating over all documents in: https://github.com/globality-corp/gaston/blob/develop/setup.py#L42 and optionally exporting documents to a file
- a cli for loading documents into a cluster from a file, we have one here: https://github.com/globality-corp/gaston/blob/develop/setup.py#L38
- a reindex cli for copying data from one index on a cluster to a new index. I think we have something like that here: https://github.com/globality-corp/gaston/blob/develop/setup.py#L45 but i can't remember the exact semantics of what we implemented

Comment (jessemyers) [inline microcosm_elasticsearch/indexing.py]: Noted.

**Pull request closed without merging**